plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/StandardDeployPreferencesPanel.java

```diff
@@ -316,17 +327,50 @@ private void createProjectIdSection() {
     Label projectIdLabel = new Label(this, SWT.LEAD);
     projectIdLabel.setText(Messages.getString("project"));
     projectIdLabel.setToolTipText(Messages.getString("tooltip.project.id"));
-    GridDataFactory.swtDefaults().align(SWT.BEGINNING, SWT.BEGINNING).applyTo(projectIdLabel);
-    projectSelector = new ProjectSelector(this);
-    GridDataFactory.fillDefaults().align(SWT.FILL, SWT.CENTER)
-      .grab(true, false).hint(400, 150).applyTo(projectSelector);
+    GridDataFactory.swtDefaults().align(SWT.BEGINNING, SWT.BEGINNING).span(1, 3)
+        .applyTo(projectIdLabel);
+
+    Label select = new Label(this, SWT.WRAP);
+    select.setText(Messages.getString("projectselector.selectProject"));
+    GridDataFactory.swtDefaults().align(SWT.FILL, SWT.BEGINNING)
+        .applyTo(select);
+
+    Link createNewProject = new Link(this, SWT.NONE);
+    createNewProject.setText(Messages.getString("projectselector.createproject",
+                                                CREATE_GCP_PROJECT_WITH_GAE_URL));
+    createNewProject.setToolTipText(Messages.getString("projectselector.createproject.tooltip"));
+    FontUtil.convertFontToItalic(createNewProject);
+    createNewProject.addSelectionListener(
+        new OpenUriSelectionListener(new ErrorDialogErrorHandler(getShell())));
+    GridDataFactory.swtDefaults().align(SWT.FILL, SWT.BEGINNING)
+        .applyTo(createNewProject);
+
+    Composite projectSelectorComposite = new Composite(this, SWT.NONE);
+    GridLayoutFactory.fillDefaults().numColumns(2).spacing(0, 0).applyTo(projectSelectorComposite);
+    GridDataFactory.fillDefaults().grab(true, false).applyTo(projectSelectorComposite);
+
+    projectSelector = new ProjectSelector(projectSelectorComposite);
+    GridDataFactory.fillDefaults().grab(true, false).hint(SWT.DEFAULT, 200)
+        .applyTo(projectSelector);
+
+    final Button refreshProjectsButton = new Button(projectSelectorComposite, SWT.NONE);
+    refreshProjectsButton.setImage(refreshIcon);
+    GridDataFactory.swtDefaults().align(SWT.END, SWT.BEGINNING).applyTo(refreshProjectsButton);
+    refreshProjectsButton.addSelectionListener(new SelectionAdapter() {
+      @Override
+      public void widgetSelected(SelectionEvent e) {
+        refreshProjectsForSelectedCredential();
+      }
+    });
+
     accountSelector.addSelectionListener(new Runnable() {
```

Comment:
This method is really long and complex. It might help to use non-anonymous inner classes instead.